drivers: flash: nrf: Change write_block_size parameter to 32-bits

adamkondraciuk opened the pull request:
According to NRF product specifications

Signed-off-by: Adam Kondraciuk <adam.kondraciuk@nordicsemi.no>

ioannisg reviewed (changes requested):
@nvlsianpu @anangl @kl-cruz pls, review
@adamkondraciuk could you elaborate more in detail about this change in the commit body?

carlescufi commented:
@adamkondraciuk I believe the reason we set it to 1 is that the driver emulates 1-byte block size, but I could be wrong. @nvlsianpu  is the right person to ask.

adamkondraciuk commented:
> @adamkondraciuk I believe the reason we set it to 1 is that the driver emulates 1-byte block size, but I could be wrong. @nvlsianpu is the right person to ask.
You're right, but this is not native NVMC write method. Moreover - while using this '"trick" you have to be aware of limitations (nWRITE parameter which allows you to write maximum two times  into one 32-bit chunk)

nvlsianpu reviewed (changes requested):
nRF driver emulates 1 byte write granularity using read-modify-write algorithm. This was don intentionally as some storage subsystem (NFFS for sure) requires 1 byte access,

I'm aware of drawback of this, eventually we can consider switch to the HW native write-block size as optionally via not-default Kconfig. If this is not option for you - pleas close the PR.

nvlsianpu commented:
@adamkondraciuk Will you work on this as optional patch (to upstream better)?

Laczen commented:
@adamkondraciuk, @nvlsianpu, what is the meaning of the nWRITE parameter ? Does this impose limitations on the 1 byte write access ? Is it allowable to write a 32-bit chunk in 4 writes (not changing what was already written) ? Do these writes have to be in order ? Does allowing more than 2 writes have a negative effect on life?

adamkondraciuk commented:
> @adamkondraciuk, @nvlsianpu, what is the meaning of the nWRITE parameter ? Does this impose limitations on the 1 byte write access ? Is it allowable to write a 32-bit chunk in 4 writes (not changing what was already written) ?

nWRITE tells you how many times you can write into one 32-bit location. In most cases this parameter is equal to 2, so you can't store 32-bit chunk in 4 writes. 

>  Do these writes have to be in order ?
No - there is no restrictions with write order. NVMC allows only to write '0' (while erased memory has all bits set to '1'). When you write '1' the bit remains unchanged - they has the last state. This 'trick'  is used to write less than 32-bit - if you set the bits in other bytes to '1' they will be unchanged (mask). However - nWRITE restriction still applies in this 32-bit location

> Does allowing more than 2 writes have a negative effect on life?
The product specification doesn't mention about it, however I'll investigate it and let you know

Laczen commented:
@adamkondraciuk, thanks for the reply. Based upon your remark and some discussions on nordic dev zone I would conclude that it is not safe to do the writing in byte size. Although it might work it is not sure that stored bytes are not changed by the new write.

The method used by the nordic-hal could potentially cause problems. I don't know if someone tested writing to flash one byte at a time and then validating what was written. Maybe a good test would be to write \b 01010101 until a page is full and to verify the result.

BTW the method used to write multiple times in the same 32-bit chunk has no influence on the expected life, it only influences the "ones" that would need to remain set, these cannot be guaranteed.

Laczen commented:
@adamkondraciuk, I have done a test on nrf52 with the zephyr flash api doing a one byte write for a full page and I have not been able to get errors. Do you have an idea what would be the worse byte pattern to write ?

adamkondraciuk commented:
> @adamkondraciuk, I have done a test on nrf52 with the zephyr flash api doing a one byte write for a full page and I have not been able to get errors. Do you have an idea what would be the worse byte pattern to write ?

I have no idea. But you can't trust the content of the word if you exceed nWRITE parameter on it - manufacturer doesn't give any guarantee. Even if it works fine now with your tests you can't be sure flawless operation in other conditions or with another product series.

Laczen commented:
@adamkondraciuk, @nvlsianpu, after studying the nordic hal it seems that writing at byte level is supported by the hal. The nrf flash driver in zephyr could be simplified. The remark and option would be void in this case.

nvlsianpu commented on samples/drivers/CAN/src/main.c:
```diff
@@ -1,277 +0,0 @@
-/*
```
not related file deletion.

adamkondraciuk replied:
sorry - fixed

nvlsianpu commented on drivers/flash/Kconfig.nrf:
```diff
@@ -33,4 +33,13 @@ config SOC_FLASH_NRF_UICR
 	  Enable operations on UICR. Once enabled UICR are written or read as
 	  ordinary flash memory. Erase is possible for whole UICR at once.
 
+config EMULATE_ONE_BYTE_WRITE_ACCESS
```
@anangl @ioannisg @carlescufi This is proposal for making possible to use native write-block-size of nRF flash instead of byte access. It is according to my recommendation. If you are agree on this change (it is backward compatible) then @adamkondraciuk might continue with the patch.

adamkondraciuk committed:
drivers: flash: nrfx: Change write_block_size parameter to 32-bits

According to NRF product specifications the smallest write unit is
32-bits. Added parameter 'CONFIG_EMULATE_ONE_BYTE_WRITE_ACCESS'
which allows one byte flash write using more complex operation.

Signed-off-by: Adam Kondraciuk <adam.kondraciuk@nordicsemi.no>

carlescufi commented:
@nvlsianpu I have no objections to this.

rakons commented:
I have only some doubts that it is enabled by default. It might lead to the missunderstanding that you are able, how @Laczen asked, write a word using 4 byte write accesses. If user wishes to enable it, he should know the limitations first.

Laczen commented:
The hal support of byte level write is what puzzles me, I would conclude from this that there is no limitation of byte level writes. If there is a limitation a hal change might be appropiate.

If the hal only provides word access, default to word writes and only enable the byte writes if it is required (only for nffs).

nvlsianpu commented:
@Laczen 
We know exactly what hal does, reason is that it is useful (I had talk with hal developers the are aware of the limitation).

@adamkondraciuk  Looks like most of people preference is to make word aligned access the default one.

Laczen commented:
> @Laczen
> We know exactly what hal does, reason is that it is useful (I had talk with hal developers, they are aware of the limitation).

This is bad, nothing is limiting the flash writing method to comply with the specifications. What if a write fails in a safety device because of the limitations ? This is bad design and should be corrected:
a. The HAL should give errors when writing to the same 32 bit address.
b. The zephyr flash interface could ignore these errors and create byte access under errors, then at least you know that you are working in unsupported mode.

rakons commented:
@Laczen 
> The HAL should give errors when writing to the same 32 bit address.

There is some point in such idea - could you propose some detailed solution for such implementation?

Laczen commented:
> @Laczen
> 
> The HAL should give errors when writing to the same 32 bit address.
> 
> There is some point in such idea - could you propose some detailed solution for such implementation?

It could be quite simple: Read to see if it has not been written and return error if it has. So the full solution would be:

a. Check if the address is a word address (and in range), return error if its not
b. Check if data is word multiple size (and in range), return error if it's not
c. Check if addresses have been written before (not all 0xffffffff), return error
d. If everything ok, do the word write.

nvlsianpu commented:
@Laczen 
Specification guarantees writing twice to the same word location without lacking of FLASH memory guaranteed endurance. This mean that this was minimal requirement when NVMC HW an FLASH itslef were designed in silicon - it not mean that more writes to the same location doesn't work (mean that memory endurance is not quarantined if write numbers are exceeded, but very probably endurance will be keep closely to the specification). This mean read-modify-write operation is allowed for nRF flash (and this was one of requirement for silicon design). That is why it is implemented in the code as well.

nvlsianpu commented:
```
/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/drivers/flash/soc_flash_nrf.c:504:8: error: unused variable 'count' [-Werror=unused-variable]
  u32_t count;
        ^~~~~
cc1: all warnings being treated as errors
```

@adamkondraciuk please fix shippable

nvlsianpu commented on drivers/flash/Kconfig.nrf:
```diff
@@ -33,4 +33,13 @@ config SOC_FLASH_NRF_UICR
+config EMULATE_ONE_BYTE_WRITE_ACCESS
+	bool "8-bit write block size emulation"
+	depends on SOC_FLASH_NRF
+	default n
```
I wonder whether not better makes this 'y' which keep default behavior as before.
Lets have a pool here: @rakons @pdunaj @ioannisg @lemrey @carlescufi ?

carlescufi replied:
I would say no. It adds code, so let's default to no.
@adamkondraciuk please remove the `default n` line, Kconfig bools are `n` by default.

fnde-ot replied:
Agreed with @carlescufi, however is this considered to be an API change?

nvlsianpu replied:
ok, so agreed. I know that @rakons has the same thought on that.

fnde-ot commented on drivers/flash/Kconfig.nrf:
```diff
@@ -33,4 +33,13 @@ config SOC_FLASH_NRF_UICR
+	default n
+	help
+	  When this option is enabled writing chunks less than minimal write
+	  block size parameter (imposed by manufacturer) is possible but opration
```
opration -> operation

fnde-ot commented on drivers/flash/soc_flash_nrf.c:
```diff
@@ -158,6 +158,12 @@ static int flash_nrf_write(struct device *dev, off_t addr,
 		return -EINVAL;
 	}
 
+#if !defined(CONFIG_EMULATE_ONE_BYTE_WRITE_ACCESS)
```
Maybe the IS_ENABLED macro should be used instead?

fnde-ot commented on drivers/flash/soc_flash_nrf.c:
```diff
@@ -503,10 +512,11 @@ static int write_op(void *context)
 	}
 #endif /* CONFIG_SOC_FLASH_NRF_RADIO_SYNC */
-
+#if defined(CONFIG_EMULATE_ONE_BYTE_WRITE_ACCESS)
```
Maybe the IS_ENABLED macro should be used instead?

fnde-ot commented on drivers/flash/soc_flash_nrf.c:
```diff
@@ -553,7 +563,7 @@ static int write_op(void *context)
 #endif /* CONFIG_SOC_FLASH_NRF_RADIO_SYNC */
 	}
-
+#if defined(CONFIG_EMULATE_ONE_BYTE_WRITE_ACCESS)
```
Maybe the IS_ENABLED macro should be used instead?

nvlsianpu commented:
@ioannisg @fnde-ot @carlescufi Please revisit.

fnde-ot reviewed (approved):
I was more thinking of removing preprocessor #if and using IS_ENABLED in your C code, but it is not that important unless someone else objects.

carlescufi commented:
>Agreed with @carlescufi, however is this considered to be an API change?

@fnde-ot no, because we are not changing the common API semantics, just a particular implementation.

nvlsianpu commented:
> @nvlsianpu please add a note to the release notes since the nRF behavior has changed by default.

done.

nvlsianpu commented:
@carlescufi Please revisit in your convenience.
